Tools/autotest/arduplane.py

```diff
@@ -104,404 +104,421 @@ def init(self):
+            raise ErrorException()
+        except ErrorException as ex:
+            self.progress("wait_level_flight failed with %s" % type(ex).__name__)
+            raise ErrorException()
 
     def change_altitude(self, altitude, accuracy=30):
         """Get to a given altitude."""
-        self.mavproxy.send('mode FBWA\n')
-        self.wait_mode('FBWA')
-        alt_error = self.mav.messages['VFR_HUD'].alt - altitude
-        if alt_error > 0:
-            self.set_rc(2, 2000)
-        else:
-            self.set_rc(2, 1000)
-        if not self.wait_altitude(altitude-accuracy/2, altitude+accuracy/2):
+        try:
+            self.mavproxy.send('mode FBWA\n')
+            self.wait_mode('FBWA')
+            alt_error = self.mav.messages['VFR_HUD'].alt - altitude
+            if alt_error > 0:
+                self.set_rc(2, 2000)
+            else:
+                self.set_rc(2, 1000)
+            self.wait_altitude(altitude-accuracy/2, altitude+accuracy/2)
+            self.set_rc(2, 1500)
+            self.progress("Reached target altitude at %u" %
+                          self.mav.messages['VFR_HUD'].alt)
+            return self.wait_level_flight()
+        except ErrorException as ex:
+            self.progress("change_altitude failed with %s" % type(ex).__name__)
             return False
-        self.set_rc(2, 1500)
-        self.progress("Reached target altitude at %u" %
-                      self.mav.messages['VFR_HUD'].alt)
-        return self.wait_level_flight()
 
     def axial_left_roll(self, count=1):
         """Fly a left axial roll."""
         # full throttle!
-        self.set_rc(3, 2000)
-        if not self.change_altitude(self.homeloc.alt+300):
+        try:
+            self.set_rc(3, 2000)
+            self.change_altitude(self.homeloc.alt+300)
+
+            # fly the roll in manual
+            self.mavproxy.send('switch 6\n')
+            self.wait_mode('MANUAL')
+
+            while count > 0:
+                self.progress("Starting roll")
+                self.set_rc(1, 1000)
+                self.wait_roll(-150, accuracy=90)
+                self.wait_roll(150, accuracy=90)
+                self.wait_roll(0, accuracy=90)
+                count -= 1
+
+            # back to FBWA
+            self.set_rc(1, 1500)
+            self.mavproxy.send('switch 4\n')
+            self.wait_mode('FBWA')
+            self.set_rc(3, 1700)
+            return self.wait_level_flight()
+        except ErrorException as ex:
+            self.progress("axial_left_roll failed with %s" % type(ex).__name__)
+            # Stablize the roll before exiting
+            self.set_rc(1, 1500)
             return False
 
-        # fly the roll in manual
-        self.mavproxy.send('switch 6\n')
-        self.wait_mode('MANUAL')
-
-        while count > 0:
-            self.progress("Starting roll")
-            self.set_rc(1, 1000)
-            if not self.wait_roll(-150, accuracy=90):
-                self.set_rc(1, 1500)
```

Comment:
@peterbarker On failure, the `set_rc(1,1500)` will be executed when an exception occurs. But this is more generalised. Any suggestions or is it already OK?

Reply:
*On line 327